esda/shape.py

```diff
@@ -0,0 +1,376 @@
+import pygeos
+import geopandas, pandas
+import numpy
+from numba import njit, prange
+
+# -------------------- UTILITIES --------------------#
+def _cast(collection):
+    """
+    Cast a collection to a pygeos geometry array.
+    """
+    if isinstance(collection, (geopandas.GeoSeries, geopandas.GeoDataFrame)):
+        return collection.values.data.squeeze()
+    elif pygeos.is_geometry(collection).all():
+        return collection
+    return pygeos.from_shapely(collection).squeeze()
+
+
+def get_angles(collection, return_indices=False):
+    """
+    Get the angles pertaining to each vertex of a set of polygons.
+    This assumes the input are polygons.
+
+    Arguments
+    ---------
+    ga  :   pygeos geometry array
+        array of polygons/multipolygons
+    return_indices  :   bool (Default: False)
+        whether to return the indices relating each geometry to a polygon
+
+    Returns
+    -------
+    angles between triples of points on each geometry, as well as the indices
+    relating angles to input geometries (if requested).
+
+    See the Notes for information on the shape of angles and indices.
+
+    Notes
+    -------
+    If a geometry has n coordinates and k parts, the array will be n - k.
+    If each geometry has n_i coordinates, then let N be a vector storing
+    those counts (computed, for example, using pygeos.get_num_coordinates(ga)).
+    Likewise, let K be a vector storing the number of parts each geometry has, k_i
+    (computed, for example, using pygeos.get_num_geometries(ga))
+
+    Then, the output is of shape (N - K).sum()
+
+    """
+    ga = _cast(collection)
+    exploded = pygeos.get_parts(ga)
+    coords = pygeos.get_coordinates(exploded)
+    n_coords_per_geom = pygeos.get_num_coordinates(exploded)
+    n_parts_per_geom = pygeos.get_num_geometries(exploded)
+    angles = numpy.asarray(_get_angles(coords, n_coords_per_geom, n_parts_per_geom))
+    if return_indices:
+        return angles, numpy.repeat(
+            numpy.arange(len(ga)),
+            pygeos.get_num_coordinates(ga) - pygeos.get_num_geometries(ga),
+        )
+    else:
+        return angles
+
+
+@njit
+def _get_angles(points, n_coords_per_geom, n_parts_per_geom):
```

Comment:
On input, `points` is an array containing the points corresponding to each shape. Since shapes in GEOS are "closed," meaning they duplicate their first and last point, this has one more point per geometry than needed. 
`n_coords_per_geom` tells us how many coordinates per geometry there are, including the closure point. Each geometry is guaranteed to be simple/single-part from the input. 

`n_parts_per_geom` is unused, and will be removed in a second. 

Say a geometry has k coordinates, including the closure point. The algorithm iterates over triples of points for each geometry. The triples are "left-aligned," so that the first triple is (0,1,2,), and the angle being measured is (1,0),(1,2). Because of the closure point, the left coordinate (k) and middle coordinate (0) would be the same. So, we need to avoid computing that tuple, and move to the next geometry. To do this, once the left index hits the kth point for the geometry, the algorithm advances the "offset" by the number of coordinates in the current geometry, as well as incrementing the index for the geometry itself. 

The issue was that I was advancing the offset by one less than necessary. Since `on_coord` is always one less than the starting index for the next geometry, you need to increment the offset by `on_coord + 1`, not `on_coord`. The existing implementation slowly gets out of sync with the point array.